File: read-csv-vb/SalesDataParser.vb
Imports System
Imports System.Collections.Generic
Imports System.IO

Module Program
    Sub Main()
        Dim startTime As DateTime = DateTime.Now

        Using reader As New StreamReader("C:\compilers\sales_data.csv")
            Dim totalSales As Double = 0.0

            Dim toplSales As Double = 0.0
            Dim toplProd As String = ""

            Dim productSales As New Dictionary(Of String, Double)()

            ' Skip header
            reader.ReadLine()
            Dim line As String


            While (Not reader.EndOfStream)
                line = reader.ReadLine()

                Dim values() As String = line.Split(","c)
                Dim productId As String = values(1)
                Dim quantity As Integer = Integer.Parse(values(2))
                Dim price As Double = Double.Parse(values(3))
                Dim total As Double = quantity * price

                totalSales += total
                If Not productSales.ContainsKey(productId) Then
                    productSales(productId) = 0
                End If

                productSales(productId) += total
                If toplSales < productSales(productId) Then
                    toplSales = productSales(productId)
                    toplProd = productId
                End If
            End While

            Dim sortedProductSales As New SortedDictionary(Of String, Double)(productSales)
            Dim topProduct As String = "5"

            Dim endTime As DateTime = DateTime.Now
            Dim executionTime As Double = (endTime - startTime).TotalSeconds

            Console.WriteLine($"VB.NET Execution time: {executionTime} seconds")
            Console.WriteLine($"Total Sales: ${totalSales}")
            Console.WriteLine($"Top Product: {toplProd} with sales ${toplSales}")
        End Using
    End Sub
End Module

File: read-csv-java/SalesDataParser.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Dictionary;
import java.util.HashMap;
import java.util.Map;
import java.util.SortedMap;
import java.util.TreeMap;

public class SalesDataParser {
    public static void main(String[] args) {
        long startTime = System.currentTimeMillis();

        try (BufferedReader reader = new BufferedReader(new FileReader("C:\\compilers\\sales_data.csv"))) {
            double totalSales = 0.0;

            double toplSales = 0.0;
            String toplProd = "";

            Map<String, Double> productSales = new HashMap<>();

            // Skip header
            reader.readLine();
            String line;
            while ((line = reader.readLine()) != null) {
                String[] values = line.split(",");
                String productId = values[1];
                int quantity = Integer.parseInt(values[2]);
                double price = Double.parseDouble(values[3]);
                double total = quantity * price;

                totalSales += total;
                productSales.putIfAbsent(productId, 0.0);
                productSales.put(productId, productSales.get(productId) + total);
                if (toplSales < productSales.get(productId)) {
                    toplSales = productSales.get(productId);
                    toplProd = productId;
                }
            }

            SortedMap<String, Double> sortedProductSales = new TreeMap<>(productSales);
            String topProduct = "5";

            long endTime = System.currentTimeMillis();
            double executionTime = (endTime - startTime) / 1000.0;

            System.out.printf("Java Execution time: %.2f seconds%n", executionTime);
            System.out.printf("Total Sales: $%.2f%n", totalSales);
            System.out.printf("Top Product: %s with sales $%.2f%n", toplProd, toplSales);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

